### PR TITLE
Added command options to define the dev entry point password

### PR DESCRIPTION
**Pull request by aschempp**
fixes https://github.com/contao/manager-bundle/issues/23

Symfony 3.3 introduced the .env component (http://symfony.com/blog/new-in-symfony-3-3-dotenv-component). The app_dev.php password is always stored there, which has multiple advantages:
 1. @discordier can always use that variable as requested in #23 
 2. the file is persistent across multiple `composer:update` commands (that will re-run `contao:install-web-dir` and would delete the password
 3. The .env file can be deployed to dev but not to prod

### CODE REVIEW

**Review by Toflar — APPROVED** [review]

File: src/Command/InstallWebDirCommand.php [hunk 1]
```diff
@@ -58,22 +71,77 @@ protected function configure()
...
+
+        if (true === $input->getOption('no-dev') && (null !== $user || null !== $password)) {
+            throw new \InvalidArgumentException('Cannot set a password in no-dev mode!');
+        }
+
+        if (null === $password && null !== $user) {
+            throw new \InvalidArgumentException('Cannot set a username without password.');
+        }
+
+        if (true !== $password) {
```
Comment (ausi) [inline src/Command/InstallWebDirCommand.php:111]: If I call `contao:install-web-dir --password=foo` the method would `return` here but the `user` option would still be missing?

Comment (aschempp) [reply src/Command/InstallWebDirCommand.php:111]: That is correct. The idea was to only ask for user and password if `--password` is used as an option. Both values are validated in https://github.com/contao/manager-bundle/pull/33/files#diff-0211dd218189f49a55922c37e4404c66R211
Otherwise we would also need `--user` as an option, or what if you use `--user=foo --password`. That would all not really make sense.

**Review by leofeyer — CHANGES_REQUESTED** [review]
I would prefer if there was only an `--add-credentials` (or similar) option and username and password were always asked interactively. It is bad practice to provide passwords in the command like `--password=clear-text`, because other users might be able to se the clear-text password in the process list!

**Comment by aschempp** [comment]
That is exactly the same discussion we had with the `UserPasswordCommand` and we decided to add the non-interactive option…

**Comment by leofeyer** [comment]
Yes, but you are now using `--password` to enable the dialog and `--password=value` to provide the password with the command. That's confusing, isn't it?

**Comment by aschempp** [comment]
That's similar to what different unix tools do, e.g. `mysql -p` to ask for password or `mysql -p foobar` to pass the password…

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Well played. 😄

File: src/Command/InstallWebDirCommand.php [hunk 2]
```diff
@@ -108,8 +181,75 @@ private function addFiles($webDir)
...
+
+        if (true === $input->getOption('no-dev') && (null !== $user || null !== $password)) {
+            throw new \InvalidArgumentException('Cannot set a password in no-dev mode!');
+        }
+
+        if ((null === $password && null !== $user) || (null === $user && null !== $password)) {
+            throw new \InvalidArgumentException('Must have username and password to set the access key.');
+        }
+
+        $accessKey = hash(
```
Comment (discordier) [inline src/Command/InstallWebDirCommand.php]: IMO we should use the proper security function here:
```diff
-        $accessKey = hash(
-            'sha512',
+        $accessKey = password_hash(
             $input->getOption('user').':'.$input->getOption('password'),
+            PASSWORD_BCRYPT,
+            ['cost' => 12]
         );
```

File: src/Resources/web/app_dev.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    || isset($_SERVER['HTTP_X_FORWARDED_FOR'])
+    || !(in_array(@$_SERVER['REMOTE_ADDR'], ['127.0.0.1', 'fe80::1', '::1']) || php_sapi_name() === 'cli-server')
+) {
+    if (false === $accessKey) {
+        header('HTTP/1.0 403 Forbidden');
+        die(sprintf('You are not allowed to access this file. Check %s for more information.', basename(__FILE__)));
+    }
+
+    if (!isset($_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW'])
+        || hash('sha512', $_SERVER['PHP_AUTH_USER'].':'.$_SERVER['PHP_AUTH_PW']) !== $accessKey
```
Comment (discordier) [inline src/Resources/web/app_dev.php]: IMO we should use the proper security function here:
```diff
-        || hash('sha512', $_SERVER['PHP_AUTH_USER'].':'.$_SERVER['PHP_AUTH_PW']) !== $accessKey
+        || password_verify($_SERVER['PHP_AUTH_USER'].':'.$_SERVER['PHP_AUTH_PW'], $accessKey)
```

File: src/Resources/web/app_dev.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ */
+
+use Contao\ManagerBundle\ContaoManager\Plugin as ManagerBundlePlugin;
+use Contao\ManagerBundle\HttpKernel\ContaoKernel;
+use Doctrine\Common\Annotations\AnnotationRegistry;
+use Symfony\Component\Debug\Debug;
+use Symfony\Component\Dotenv\Dotenv;
+use Symfony\Component\HttpFoundation\Request;
+
+// The check is to ensure we don't use .env in production
```
Comment (discordier) [inline src/Resources/web/app_dev.php]: While I do not really get what you mean with this comment (the check is in `app_dev.php` and has nothing to do with `app.php`), it feels wrong to me with respect to future compat with symfony flex and also with current implementation.
This causes not only this feature to only work in dev mode but also all other `getenv` stuff.
IMO we should use `dotenv` in both endpoints, `app.php` and `app_dev.php` to make behaviour consistent.
You are right however about not adding the auth check to `app.php`.

So please also add this also to `app.php`

**Review by discordier — CHANGES_REQUESTED** [review]

File: src/Command/InstallWebDirCommand.php (hunk 2 above)
Comment (discordier) [reply src/Command/InstallWebDirCommand.php]: The cost is of course just an example

Comment (aschempp) [reply src/Command/InstallWebDirCommand.php]: This was copied from the current standard-edition, not sure if we want to change that. @contao/developers

File: src/Resources/web/app_dev.php (hunk 4 above)
Comment (aschempp) [reply src/Resources/web/app_dev.php]: https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/3.3/web/index.php#L10

`.env` is meant to load/override environment variables in development or local environment, not in production. Production should use actual environment variables. Also, as you pointed out, we don't need anything from `.env` in prod, so no reason to read and parse a unused file.

However, the comment sure is wrong 😉

Comment (backbone87) [reply src/Resources/web/app_dev.php]: in symfony flex there is only a single front controller for all envs, it configures itself from env vars. the .env file gets loaded, if there is no APP_ENV var set already

File: tests/Command/InstallWebDirCommandTest.php [hunk 5]
```diff
@@ -72,7 +76,7 @@ public function setUp()
      */
     public function tearDown()
     {
-        $this->filesystem->remove($this->tmpdir);
+        (new Filesystem())->remove($this->tmpdir);
```
Comment (backbone87) [inline tests/Command/InstallWebDirCommandTest.php:80]: why this change? there is a filesystem in `$this->filesystem`

File: src/Resources/web/app_dev.php (hunk 4 above)
Comment (discordier) [reply src/Resources/web/app_dev.php]: @aschempp I have to second the statement by @backbone87.

The original line from symfony (`if (!getenv('APP_ENV')) {`) has absolutely nothing in common with your `file_exists` check here.
The whole workflow is changed.
IMO we should keep the behaviour of symfony and therefore add the `dotenv` to both endpoints.

> Also, as you pointed out, we don't need anything from .env in prod, so no reason to read and parse a unused file.

You mean, your endpoint does not need anything so far but can you guarantee that no bundle does some `getenv` because ist is already `dotenv` aware?
What do you expect users to do when they want to activate it? I mean, the can not alter the endpoint at all as the cm will overwrite it over and over again.

Comment (aschempp) [reply src/Resources/web/app_dev.php]: > can you guarantee that no bundle does some getenv because ist is already dotenv aware?
> What do you expect users to do when they want to activate it?

There is no *dotenv aware*. Dotenv is to configure your environment, not extensions or bundles. That's what your Symfony config is for, in our case the parameters.yml. The only way an env variable could be used is by placing such a configuration in parameters.yml, and that means you know about environment variables and should use them on your server platform.

There is no case where a **managed edition** has variables it does not know of. If you have that case, you either do not have a managed edition. Or you can always not generate the entry points and adjust them to your needs.

File: tests/Command/InstallWebDirCommandTest.php (hunk 5 above)
Comment (aschempp) [reply tests/Command/InstallWebDirCommandTest.php:80]: It was not available for unknown reasons… maybe something changed in phpunit 5?

File: src/Resources/web/app_dev.php (hunk 4 above)
Comment (backbone87) [reply src/Resources/web/app_dev.php]: indeed no bundle should directly access env vars, but use the DI containers parameters.
however there is no parameters.yml in symfony flex anymore, since it is expected that all deployment target dependant config is passed through env vars and then passed through to the containers parameters. how this applies to contao settings like "allowed file types" idk, because that is clearly not a config you would pass through env vars. imho the cleanest way, would be to use symfonies abilities to stack semantic config and allow to a contao admin to edit the semantic config directly (and not through container parameters).

Comment (aschempp) [reply src/Resources/web/app_dev.php]: We're not on Symfony flex, so this does not really apply to us…

File: src/Command/InstallWebDirCommand.php (hunk 2 above)
Comment (discordier) [reply src/Command/InstallWebDirCommand.php]: The standard edition clearly states:
> // This check prevents access to debug front controllers that are deployed by accident to production servers.
> // Feel free to remove this, extend it, or make something more sophisticated.

And as we always deploy the entry point and not only "by accident" but on purpose, we IMO definitely should "make something more sophisticated".

**Comment by aschempp** [comment]
We already do? We have basic authentication, the standard does not?

File: src/Command/InstallWebDirCommand.php (hunk 2 above)
Comment (discordier) [reply src/Command/InstallWebDirCommand.php]: Reply to https://github.com/contao/manager-bundle/pull/33#issuecomment-304280979 which should have been here but got sorted on the generic discussion below:
> We already do? We have basic authentication, the standard does not?

Let me put it another way, we have strong authentication (with bcrypt, salting etc.) in application for Backend and Frontend users, we have it for the manager itself (uses bcrypt as suggested by me here) but the dev entry point shall be protected by sha512 only?
Given that most people (IMO) will add an generic user named "admin" or "developer" this will open us up for rainbow table attacks.

I wonder why nobody else from @contao/developers has an opinion on this one.

File: src/Resources/web/app_dev.php (hunk 4 above)
Comment (discordier) [reply src/Resources/web/app_dev.php]: But bundles are shifting over and generate the container config from env vars which is the way symfony wants to go.
So let's be an early adopter of the concept in general (which does not hurt at all).

I also want to stress out that the need for "dotenv" is higher on deployment than on dev machines because you usually have more control over the dev machine than over the deployment target (you can not set an env var on shared hosting apache that easy).

It will still produce side effects as simply grep'ing for "getenv" in my local vendor directory of a plain symfony installation gives many calls.

I simply fail to see why removing a cool feature from the non dev edition will do any harm when the dev edition shall benefit from it but am willing to get convinced if anyone provides good reasons.

File: src/Command/InstallWebDirCommand.php [hunk 6]
```diff
@@ -108,8 +181,75 @@ private function addFiles($webDir)
...
+            throw new \InvalidArgumentException('Cannot set a password in no-dev mode!');
+        }
+
+        if ((null === $password && null !== $user) || (null === $user && null !== $password)) {
+            throw new \InvalidArgumentException('Must have username and password to set the access key.');
+        }
+
+        $accessKey = password_hash(
+            $input->getOption('user').':'.$input->getOption('password'),
+            PASSWORD_DEFAULT
```
Comment (leofeyer) [inline src/Command/InstallWebDirCommand.php]: This should be `PASSWORD_BCRYPT`, shouldn't it?

**Review by leofeyer — CHANGES_REQUESTED** [review]
Why are the changes from https://github.com/contao/manager-bundle/commit/4bdf78528127f1de7e90911fb31b12b73e96c37f in the PR now?

**Comment by aschempp** [comment]
What are you missing?

**Comment by leofeyer** [comment]
I'm not missing anything, there is too much in the PR. 😄 Can you please rebase?

**Comment by leofeyer** [comment]
Never mind. The PR was still targeted against the `develop` branch.
